pkg/bpf/exporter.go

```diff
@@ -192,38 +193,221 @@ func (e *exporter) Detach() {
+
+		select {
+		case <-stopChan:
+			if err := rd.Close(); err != nil {
+				return fmt.Errorf("closing ring buffer reader: %v", err)
+			}
+			return nil
+		default:
+			var event keplerEvent
+			record = new(ringbuf.Record)
+
+			err := rd.ReadInto(record)
+			if err != nil {
+				if errors.Is(err, ringbuf.ErrClosed) {
+					return nil
+				}
+				if errors.Is(err, ringbuf.ErrFlushed) {
+					record.RawSample = record.RawSample[:0]
+				}
+				klog.Errorf("reading from reader: %s", err)
+				continue
+			}
+
+			if err := binary.Read(bytes.NewBuffer(record.RawSample), binary.LittleEndian, &event); err != nil {
+				klog.Errorf("parsing ringbuf event: %s", err)
+				continue
+			}
+
+			// Process the event
+			e.handleEvent(event)
 		}
-		if err != nil {
-			return nil, fmt.Errorf("failed to batch lookup and delete: %v", err)
+	}
+}
+
+type bpfMetrics struct {
+	CGroupID        uint64
+	CPUCyles        PerCPUCounter
+	CPUInstructions PerCPUCounter
+	CacheMiss       PerCPUCounter
+	CPUTime         PerCPUCounter
+	TxIRQ           uint64
+	RxIRQ           uint64
+	BlockIRQ        uint64
+	PageCacheHit    uint64
+}
+
+func (p *bpfMetrics) Reset() {
+	p.CPUCyles.Reset()
+	p.CPUInstructions.Reset()
+	p.CacheMiss.Reset()
+	p.CPUTime.Reset()
+	p.TxIRQ = 0
+	p.RxIRQ = 0
+	p.BlockIRQ = 0
+	p.PageCacheHit = 0
+}
+
+func newBpfMetrics() *bpfMetrics {
+	return &bpfMetrics{
+		CPUCyles:        NewPerCPUCounter(),
+		CPUInstructions: NewPerCPUCounter(),
+		CacheMiss:       NewPerCPUCounter(),
+		CPUTime:         NewPerCPUCounter(),
+	}
+}
+
+type PerCPUCounter struct {
+	Values map[uint64]uint64
+	Total  uint64
+}
+
+func NewPerCPUCounter() PerCPUCounter {
+	return PerCPUCounter{
+		Values: make(map[uint64]uint64),
+	}
+}
+
+func (p *PerCPUCounter) Start(cpu, taskID uint32, value uint64) error {
+	key := uint64(cpu)<<32 | uint64(taskID)
+
+	// TODO: The eBPF code would blindly overwrite the value if it already exists.
+	// We will preserve the old behavior for now, but we should consider
+	// returning an error if the value already exists.
+	p.Values[key] = value
+
+	return nil
+}
+
+func (p *PerCPUCounter) Stop(cpu, taskID uint32, value uint64) error {
+	key := uint64(cpu)<<32 | uint64(taskID)
+
+	if _, ok := p.Values[key]; !ok {
+		return nil
+	}
+
+	if value == 0 {
+		return nil
+	}
+
+	var delta uint64
+
+	// Probably a clock issue where the recorded on-CPU event had a
+	// timestamp later than the recorded off-CPU event, or vice versa.
+	if value > p.Values[key] {
+		delta = value - p.Values[key]
+	} else {
+		delta = 0
+	}
+	p.Total += delta
+
+	delete(p.Values, key)
+	return nil
+}
+
+func (p *PerCPUCounter) Reset() {
+	// Leave values in place since we may have in-flight
+	p.Total = 0
+}
+
+func (e *exporter) handleEvent(event keplerEvent) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	var p *bpfMetrics
+
+	if _, ok := e.processMetrics[event.Pid]; !ok {
+		e.processMetrics[event.Pid] = newBpfMetrics()
+	}
+	p = e.processMetrics[event.Pid]
```

Comment:
We can extract better performance by locking only read and write to data. 
Have you considered a RWMutex? 

Is the purpose of Mutex to lock the entirely of the exporter or only the processMetrics map?

Reply:
> Is the purpose of Mutex to lock the entirely of the exporter or only the processMetrics map?

The idea is to lock the entire exporter - so RWMutex won't help.
We want to stop reading new updates from the RingBuffer until after `CollectProcesses` has exited and all the `Total` counters have been reset back to 0.

Reply:
Thanks for the clarification.  Would you mind moving the mutex to the top of struct ? 
You can find similar patterns in go . The idea is to keep mutex above what it protects.

Reply:
I've kept it where it was and added code comments. It is protecting processMetrics but also freedPIDs.